Chore: Use conventional changelog and automate config version bumps

MisterTimn opened the pull request:
#### 📁 Related issues

#145 

#### ✍️ Description

This PR tries to automate and streamline some of the release procedure described in `docs/release.md`.

- `pre-release.js` has been added, it is called along with `npm version` and bumps versions of @solid/community-server references in the configs and `package.json`
- [`conventional-changelog`](https://www.npmjs.com/package/conventional-changelog) has been added as a dev-dependency and is called  with `npm version` in favour of `manual-git-changelog`.

**Todo**
- [x] `package.json` changes are all commited in the `chore: config update to version vx.0.0` commit, so not in the release commit. Either use pre/post or some workaround.
- [x] don't use `shellJs` 😢 
- [x] decide where this script lives

### ✅ PR check list

Before this pull request can be merged, a core maintainer will check whether
  * [x] this PR is labeled with the correct semver label
    - semver.patch: Backwards compatible bug fixes.
    - semver.minor: Backwards compatible feature additions.
    - semver.major: Breaking changes. This includes changing interfaces or configuration behaviour.
  * [ ] the correct branch is targeted. Patch updates can target main, other changes should target the latest versions/* branch.
  * [ ] the RELEASE_NOTES.md document in case of relevant feature or config changes.
  * [ ] any relevant documentation was updated to reflect the changes in this PR.

<!-- Try to check these to the best of your abilities before opening the PR -->

MisterTimn commented:
https://github.com/CommunitySolidServer/CommunitySolidServer/issues/145#issuecomment-1119561377
As discussed in the related issue, another option would be to contribute to `componentsjs` and add a bin script that can be called from it. This scripts sole purpose would be to detect and update its references.

So either I move forward with this PR and the script, then later potentially contribute it to `componentsjs`.

RubenVerborgh commented:
`/scripts` would be good indeed; could move some other stuff there too.

MisterTimn commented:
Output of `npm version` right now:

![image](https://user-images.githubusercontent.com/4209558/167376060-28c4d67a-5a44-42ac-b9b9-21ec9f96a622.png)

`changelog.md` output for the alpha update as an example:
# [5.0.0-alpha.0](https://github.com/CommunitySolidServer/CommunitySolidServer/compare/v4.0.0...v5.0.0-alpha.0) (2022-05-05)

Still have to look into config options for this tool, might be able to more closely resemble our standard output

### Bug Fixes

* %2F not handled correctly in file backend [#1184](https://github.com/CommunitySolidServer/CommunitySolidServer/issues/1184) ([dbdb9b4](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/dbdb9b424e4c4f68c19c66396064486bff93a7e4))
* Always define [@type](https://github.com/type) in configs ([cfdd122](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/cfdd1221f05f2b51abc770e64407bb8392224383))
* Change YargsCliExtractor structure to avoid Components.js issues ([6f4e70d](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/6f4e70dbb928a9cb64e83a71954ea607a0bdb1a0))
* **deps:** Fix missing dependency from package-lock.json ([f2c9c0c](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/f2c9c0cb78fa89d018409f4eae51329289be795e))
* **deps:** Update to Comunica v2 ([1de1f7c](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/1de1f7c12a8ae3f89fd91e9cb33af0405af7d995))
* Fix typing issues with latest Components.js generator ([0e32d9a](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/0e32d9ad9b3e8558c32bf7263e33af157b88a6a1))
* Make delimiter encoding case-insensitive. ([50469e2](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/50469e2c1f3d9c808062fde96d2ce62d5e85475e))
* Update prefixes in all configs ([ce27bec](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/ce27bec207652f0bd37beb23301f864464751438))


### Features

* add a process-/thread-safe file-based ResourceLocker ([fa78bc6](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/fa78bc68567d836927731eaa9a7d7f752dadae49))
* file-based backend fallback for unknown media types ([ff80079](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/ff80079000a0f308f31b83cfc52e90421f20dadc))

MisterTimn commented:
@joachimvh The changelog formatting is quite different, all included presets of conventional changelog follow similar styling (strip the prefix, add a commit link in brackets). They mostly differ in Title names derived from prefixes.

Do we adhere to the convention more or keep things as is? If we want to keep it I can either
- remove the `conventional-changelog` and keep using `manual-git-changelog` (maybe its the better tool for the job in this case)
- make a custom preset (based off of https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-conventionalcommits) for our styling
- maybe add some automated postformatting

MisterTimn commented:
> I'm assuming simply using the `conventional-changelog` format would be the easiest as that requires the least effort from our side?

Yes that would indeed be easiest. I will change the existing changelog to the conventional format and include it in this PR. But I would only change the bullet points, not the titles because we use the **Added**, **Changed** and **Deprecated** keyword to subdivide (mostly) features. Found option to change the subtitles and auto hide certain prefixes if wanted

```json
{
    "options": {
        "preset": {
            "name": "conventionalcommits",
            "types": [
                {"type": "feat", "section": "Features"},
                {"type": "fix", "section": "Fixes"},
                {"type": "chore", "hidden": "Chores"},
                {"type": "docs", "hidden": "Docs"},
                {"type": "style", "hidden": true},
                {"type": "refactor", "hidden": true},
                {"type": "perf", "hidden": true},
                {"type": "test", "hidden": true}
            ],
            "header": "Changelog"
        }
    }
}
```

joachimvh commented:
Would not auto-hide prefixes (yet?) as sometimes it might make sense to include them.

MisterTimn commented:
> Would not auto-hide prefixes (yet?) as sometimes it might make sense to include them.

Yes I'll just add them with a similar name (the default of the preset hides all except `feat` and `fix`)

Markdown re-generated for 4.0.0 as example below
___
## [4.0.0](https://github.com/CommunitySolidServer/CommunitySolidServer/compare/v3.0.0...v4.0.0) (2022-04-19)


### Features

* add a Redis based Read-Write Locker ([e2e2d08](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/e2e2d0864f9b81c56fea61067acdc011347b2283))
* Add CachedResourceSet ([0e4d012](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/0e4d012086801108c462d14cc52c669c7e59a232))
* Add RedirectingHttpHandler ([468e11d](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/468e11d906e051a463d24ea21a7ad4db53dde5f5))
* Add support for client_credentials authentication ([2ec8fab](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/2ec8fabd54823fd40ba194247715000a181f8076))
* Add utility functions for generating error classes ([f3dedf4](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/f3dedf4e27234efd6caa8a04ad0d0318aa3ba01d))
* Allow dynamically adding CLI parameters in configs ([bedab90](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/bedab907f9c62bb2e146188bf31c82272692465b))
* Check parent ACL permissions for create/delete requests ([d908374](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/d908374364d3b3d72bf8cfe111d42634503e32fd))
* Create MetadataParser that detects JSON with Context link and throws an error ([48efc6f](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/48efc6fae180108756550dc761b65fb1b71a4018))
* Create separate storage to generate keys ([a1a6ce0](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/a1a6ce01fa245e39acd7df5866b45c754704e9b9))
* Dynamically generate Allow and Accept-* headers ([6e98c6a](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/6e98c6aae4bfc53b5a002bf8314c83add9a129e5))
* Handle OPTIONS requests in OperationHandler ([ad3edcf](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/ad3edcf1a89257344d2209edd76b91f3eed9bc82))
* Make LazyLoggerFactory buffer messages. ([238570b](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/238570b3d29d19cbda8e3ba9263f34d4a5679ff4))
* new helper functions to replace regexes [#807](https://github.com/CommunitySolidServer/CommunitySolidServer/issues/807) ([283c301](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/283c301f0833718b879cf2040fab3d70f72ebf53))
* Only accept NamedNodes as predicates for metadata ([668d0a3](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/668d0a331fb8da45b0dc8e72f05d0a47fbcedd2e))
* Parse content-type more strictly ([027e370](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/027e3707fdc161fee005d4049648806f546d09d9))
* Pass access modes to PermissionReaders ([2ae5924](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/2ae5924dde55c53c8c58e447a3e3734e993287fb))
* Remove meta parameter from logging. ([2c6167e](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/2c6167e0cb57f97d412d97ddf84f5698a881175b))
* Return 404 for read/delete requests if there is no resource ([e86e0cf](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/e86e0cf36bb69858c514352c84b5ddc1900633f7))
* Store content type parameters ([a860205](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/a8602055e67b5d23ec70f3cc0dbaee6b4235fda4))
* Store methods in MethodNotAllowedHttpError ([effc20a](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/effc20a270baa2b9305eb11a7858fa2353ab4434))
* Support seeding pods and accounts ([c8d4bfe](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/c8d4bfec39e554fc920ece99d4272cff8128f342))
* Warn about UnsecureWebSocketsProtocol. ([5c21819](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/5c218193ab7ddf8ad80122279a52f1c06f662cea))


### Refactored

* Make Logger an interface. ([3685b7c](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/3685b7c659aadf583abe3c7442427682006049e3))
* Move key/value storages to relevant configs ([30ad301](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/30ad3015f04a8ecad4e387906e0971345abbce39))
* Rename resourceExists to hasResource ([4404fa0](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/4404fa07d9238aed2afdce29e3a3d1eafed6af0b))
* Use fs-extra instead of fs to simplify file access ([fe39f97](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/fe39f97ee0550065a65c8e221a5f51908922fd37))


### Docs

* Add Helm Chart reference to the README ([3b353af](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/3b353affb1f0919fdcb66172364234eb59c2e3f6))
* Add info about pod seeding to release notes ([30011ba](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/30011ba86b84f7b3569edef4e2dba7c707a556ef))
* Add parseContentType changes ([706f0d5](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/706f0d5316f314637f1bb3403f5123c07b611726))
* Add RedirectingHttpHandler entry to release notes ([559fe17](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/559fe170dcd02d9ddfd99d193a0771d6cf781e1c))
* Document data migration actions going from v2 to v3 ([330b645](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/330b645236c38dcc3b0a0c1de68ebcb25eb437f1))
* Fix new repo folder name. ([11bc657](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/11bc657e6868140ba4f39c263e0f524b74dbe0e8))
* Integrate pod seeding documentation ([f089ffd](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/f089ffda47b0d3b682a826c8fbcc1249a9d2a442))
* Mention metadata predicate change ([4f7eec3](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/4f7eec3200c0554b5491382db72f7042dd0cbfef))
* Update and extend helm instructions ([3c32466](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/3c32466d8873e73328f678acce1260fea02cabc9))
* Write initial user documentation ([a5a34f5](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/a5a34f5071a97ab42c9f17bfe7727f21ccd2461d))


### Testing

* Add flushPromises utility function ([16e9368](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/16e93687347abdbcb0faeb69673728d5589d7c7a))
* Add integration test that simulates running 2 hours ([ce5a98b](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/ce5a98b16fb6dfa54a0f69aca1c43433de8aadd3))
* Add pod seeding integration test ([db906ae](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/db906ae872ba81e8cf39f4f73494370fddbec58a))
* Create permission table to automate tests ([6f83ac5](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/6f83ac5ead5bb98fae9c09d1996b4b5c2ce2fa51))
* Update CTH run to use client credentials ([c0bc150](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/c0bc1501228c80ae8e2deb403fd2009124a96369))


### Fixes

* Add IANA type to child metadata in FileDataAccessor ([7152897](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/7152897b89b6454142ade57e610588093bf03c5d))
* Add missing imports from default configs ([30799f6](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/30799f6667490f3105bca05784c6bbf0d38b36f3))
* Extract correct access modes from request ([9a29cc2](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/9a29cc22578c5a71667445b6b28a76c3eaaf61e4))
* Keep storage paths consistent with previous version ([570e167](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/570e167a36b8da14561680d2709190ea78e15b61))
* Prevent expired storage cleanup from crashing the server ([f08cdf7](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/f08cdf75f76d7e23b5d3277aae4a113380e245a7))
* Prevent slugs with trailing slashes for non-container resources ([5965268](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/5965268ebfdb49f1ae4da3809c913016686cee07))
* Restore eslint plugins ([5e6c42e](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/5e6c42e1f82f3474c64c8f43ef1827775ce8cfa7))
* Support entries function in JsonResourceStorage ([7654801](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/76548011f22b3939a39c7ffae09e31805450bc0a))
* Throw error when accessing URLs out of scope ([d42125a](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/d42125a91d7d4a2dde5fc280869f99e10546731f)), closes [#1148](https://github.com/CommunitySolidServer/CommunitySolidServer/issues/1148) [#1148](https://github.com/CommunitySolidServer/CommunitySolidServer/issues/1148)
* Undo authorization on OPTIONS requests ([97e600b](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/97e600bf4f1e9e397de88239395216e05e6c3b4b))


### Chores

* Add init logging to conformance tests ([847dc82](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/847dc821b86757ab3c86e85c556bd5cca82f49bc))
* Add line about updating documentation to PR template ([46132da](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/46132da8c29dc05242ae035203b478f8be325083))
* Add template for v4.0.0 release notes ([bf1afdd](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/bf1afdd6ec205dc0f86cfff87782687617b8840d))
* **deps:** update actions/checkout action to v3 ([196c529](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/196c529f6b3ec2aca853c5eb3f8bc9d2533853e0))
* **deps:** update actions/download-artifact action to v3 ([982d57e](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/982d57e1d2812925b09353445a318d067a242943))
* **deps:** update actions/setup-node action to v3 ([9f6fdab](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/9f6fdab0be335163def743a2b9dabed482e997c8))
* **deps:** update actions/upload-artifact action to v3 ([311d3e0](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/311d3e0ff013679e142dda1ac4bf78a5c50a28fb))
* **deps:** update arrayify-stream to v2 ([e4f1e2a](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/e4f1e2aa0c068b3496f54593f2d060700142bed0))
* Error on linter warnings ([5d802c6](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/5d802c6a34b8d63fb6f4e5fce4505e8117730cbf))
* Fix package-lock.json ([841ec2a](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/841ec2a3eb156278b4966143bdf5b333f3c6f87d))
* Fix typo in release notes ([207419b](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/207419bfd4573da002d3d4fd9b82ed1e1c6a8df1))
* Increase CI timeout to 15 minutes ([dd56886](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/dd568869ca2228e3987c9fd998c778e181ada2bb))
* Remove validate-components from CI ([9ac00ca](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/9ac00ca82ce942282029fd18027cae4d4fcb636b))
* replace redis library with ioredis ([4a37f8c](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/4a37f8c8798fa7415a23381114a50631edb203b6))
* Simplify test mocks. ([e4688e1](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/e4688e12a19776567e82250148832ab5439472c9))
* Update configs to v4.0.0 ([784bfc8](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/784bfc8c2128aeed4f42c18a5f38632b3c20240d))
* Update conformance tests to use v4.0.0 branch ([c15efc0](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/c15efc0b82d41f2dbf625af3e0bc416952bf742e))
* Update links for new repository ([7d3320d](https://github.com/CommunitySolidServer/CommunitySolidServer/commit/7d3320dea64f6552744787af7ead9965882c83da))

joachimvh commented:
Looks good

MisterTimn commented:
I've moved to using `conventional-changelog/standard-version` in lieu of `conventional-changelog/conventional-changelog-cli`. Effect is the same, standard-version does a bit more, seems well maintained and is a drop-in replacement for `npm version`.

I've left the `version` script in the package.json so when this PR gets merged we can still default back to the old way of releasing if needed. The `upgrade-configs.js` script is also included as a first step there so configs will also get updated accordingly.

You can either install `standard-version` globally and use it from commandline directly or you can use it through `npm run release`

```
npm version major -m "chore: Release version %s of the npm package."
```
Now becomes
```
npm run release -- -r major
# OR after npm i -g standard-version
standard-version -r major
```

Running the release step will create the `chore(release): upgrade configs to v{major}.0.0` if a major upgrade is being released, before (using a lifecycle script that happens `postbump`) and then create a second commit `chore(release): Release version {{currentTag}} of the npm package.` with version bump and updated changelog. A postprocessing step is also included which only allows h2 headers for releases in the markdown file (see https://github.com/conventional-changelog/standard-version/issues/873). 

Standard-version grabs it config from `.versionrc.json`, there the changelog header can be set, as well as the standard commit message, the different subtitles, etc.

Some commands to highlight and show the use of this script
```
# Decide version automatically from conv commits
npm run release
# Create a pre-release and decide the semantic version automatically from commits
## Release maj.min.patch.0
npm run release -- --pre-release alpha
## Release maj.min.patch.alpha0
npm run release -- --pre-release
# Enforce version bump
npm run release -- -r minor
# Perform a dry-run which will print out the steps, as well as the CHANGELOG.md changes
npm run release -- --dry-run
```

joachimvh commented on scripts/upgrade-configs.js:
```diff
@@ -0,0 +1,33 @@
+async function upgradeConfig(){
+    console.log(`Changing @solid/community-server references to ${major}.0.0\n`)
+    
+    const configs = await new fdir().withBasePath()
```
Since we simply need to find all json files in the config directory and below I would just make a small recursive function that does this.

MisterTimn commented on documentation/release.md:
```diff
@@ -3,35 +3,34 @@
- * Publish with `npm publish --tag next`.
- * Do not update the branch or anything related.
+
+* Version with `npm release -- -r major --pre-release alpha` (or `npm release -- --pre-release alpha` to autogenerate the version)
```
If you don't force a version through '-r major' the version will be decided through analysis of commit history. Fixes for patch, feats for minor and breaking changes for major.

joachimvh replied:
Would not mention it at all then, since we don't really make use of the breaking tags.

MisterTimn commented on documentation/release.md:
```diff
@@ -3,35 +3,34 @@
- * Potentially upgrade the recipes at https://github.com/CommunitySolidServer/recipes
- 
+
+* Merge `main` into `versions/x.0.0`.
```
Haha, sorry I should disable format on save for markdown files in CSS

MisterTimn commented on package.json:
```diff
@@ -60,7 +60,9 @@
-    "version": "manual-git-changelog onversion",
+    "upgrade-config": "node ./scripts/upgrade-configs.js",
+    "release": "standard-version",
+    "version": "node ./scripts/upgrade-configs.js && manual-git-changelog onversion",
```
I was gonna leave it in in case you use it for next release and absolutely hate it 😬

joachimvh replied:
I'm sure I'll love it! (And I can manually revert then if needed 😉 )

RubenVerborgh replied:
And interestingly, a test will still run on commit then, so we immediately know it works.

MisterTimn commented on scripts/changelog-format.js:
```diff
@@ -0,0 +1,11 @@
+const fs = require('fs-extra');
```
I was working on another script for other issue and was actually using TS in that one, with `ts-node` to execute it. I can definitely change these to TS. Going into this I thought doing it in TS was 'too much'. 

I'll look into how its done for the `createAccountCredentials` and do it similarly!

MisterTimn commented on scripts/changelog-format.js:
```diff
@@ -0,0 +1,11 @@
+    return fs.writeFile(filePath, result, 'utf8');
+}
+
+replaceInFile(/### \[/g, '## [', 'CHANGELOG.md')
+    .then(console.log("CHANGELOG.md formatted"))
+    .catch(error => console.error(error))
```
The standard version uses `### ` header for patch versions... It's an open issue on their repo, it doesn't really make sense to do it that way. I added it to fix that and also to potentially add more formatting later.
I linked the issue in my overview comment above

MisterTimn commented:
Requested changes implemented:

- now reading version straight from `package.json` and also using the `name` defined there to find and replace in configs
- all scripts now TS, executed with ts-node

One issue: when running `npx standard-version` now, the commit triggers husky pre-commit hook for testing, but tests fail because due to those changes. Running `npm ci` then testing results in success though.

Do I add `npm ci` to the pre-commit lifecycle hook of `standard-version`?

joachimvh commented:
> One issue: when running `npx standard-version` now, the commit triggers husky pre-commit hook for testing, but tests fail because due to those changes. Running `npm ci` then testing results in success though.


Which tests fails and why specifically? Is this because of a missing dependency or the version number/contexts changing? If yes we probably just want an `npm run build` in there somewhere.

MisterTimn commented:
> Which tests fails and why specifically? Is this because of a missing dependency or the version number/contexts changing? If yes we probably just want an `npm run build` in there somewhere.

18 test suites failed, all integration tests. `npm run build` and re-run of tests still gives lots of failures.
![image](https://user-images.githubusercontent.com/4209558/168286972-2dd95bae-1603-4bf2-9625-23c4d52c14a0.png)

Found the culprit, the integration tests have separate config files, and I am not changing these. https://github.com/CommunitySolidServer/CommunitySolidServer/tree/main/test/integration/config

joachimvh commented:
Ah right, those also have to be changed, also the configurations in `/templates/config/`.

joachimvh commented on package.json:
```diff
@@ -60,7 +61,7 @@
     "test:watch": "jest --coverageReporters none --watch test/unit",
     "validate": "componentsjs-compile-config urn:solid-server:default:Initializer -c config/default.json -f > /dev/null",
-    "version": "manual-git-changelog onversion",
+    "upgrade-config": "ts-node ./scripts/upgradeConfig.ts",
```
We should also do this for `createAccountCredentials` instead of the build that happens first now in the schedule workflow. So feel free to also update that if you're looking for things to do.

joachimvh commented on scripts/tsconfig.json:
```diff
@@ -0,0 +1,6 @@
+  "include": [
+    "."
+  ]
+}
```
Missing newline

MisterTimn replied:
And now I realize I can just enable an editor setting to add newline at the end of file on save ^^

joachimvh commented on scripts/formatChangelog.ts:
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env node
```
Since we use `ts-node` and no longer build the script we can now change this and the other one to
```suggestion
#!/usr/bin/env ts-node
```

joachimvh commented on scripts/upgradeConfig.ts:
```diff
@@ -0,0 +1,107 @@
+ * @param  string - the search string to be escaped
+ * @returns Promise with the escaped string
+ */
+async function escapeString(string: string): Promise<string> {
+  return string.replace(/[$()*+.?[\\\]^{|}]/gu, '\\$&');
+}
```
We have the `escape-string-regexp` dependency that you can use.

joachimvh commented on scripts/upgradeConfig.ts:
```diff
@@ -0,0 +1,107 @@
+interface NpmPackage {
+  version: string;
+  name: string;
+}
+/**
```
```suggestion
}

/**
```
And a few more places below.

joachimvh commented on scripts/upgradeConfig.ts:
```diff
@@ -0,0 +1,107 @@
+ * Reads and return the package.json file to extract version and name
+ * @returns Promise with the NpmPackage
+ */
+async function getNpmPackage(): Promise<NpmPackage> {
+  return JSON.parse(await readFile('package.json', 'utf8'));
+}
```
We're going to have a utility function for this soon: https://github.com/CommunitySolidServer/CommunitySolidServer/pull/1286/files#diff-b2c17cdaffd9f24e5b97bb04476db87f2b0eaff6bd2e89ff0aea097b0cb06a39R274-R280

Unfortunately that is the 5.0.0 branch. So perhaps just copy that to this branch as well to the same location and then it will get solved when merging main into 5.0.0 later.

joachimvh commented on scripts/upgradeConfig.ts:
```diff
@@ -0,0 +1,107 @@
+  console.log(`Replacing version in ${filePath}`);
+  const data = await readFile(filePath, 'utf8');
+  const escapedName = await escapeString(name);
+  const regex = new RegExp(`(${escapedName}\\/)\\^\\d+\\.\\d+\\.\\d+`, 'gmu');
```
Would let this function take a regex as input since it's going to be the same for all calls.

joachimvh replied:
Regex can also be rewritten like this to include less backslashes which might help readability.
```suggestion
  const regex = new RegExp(`(${escapedName}/)${/\^\d+\.\d+\.\d+/.source}`, 'gmu');
```

joachimvh commented on scripts/upgradeConfig.ts:
```diff
@@ -0,0 +1,107 @@
+  configs.push(...await getFilePaths('templates/config/', /.+\.json/u));
+
+  for (const config of configs) {
+    await replaceComponentVersion(config, npmPackage.name, `${major}.0.0`);
```
In practice `npmPackage.name` is probably always going to work, but to completely correct you should use the `lsd:module` field from the package.json.

joachimvh commented on scripts/upgradeConfig.ts:
```diff
@@ -0,0 +1,107 @@
+  const npmPackage = await getNpmPackage();
+  const major = npmPackage.version.split('.')[0];
+
+  console.log(`Changing @solid/community-server references to ${major}.0.0\n`);
```
Should insert the name that is being changed in case the name ever changes.

joachimvh commented on scripts/upgradeConfig.ts:
```diff
@@ -0,0 +1,107 @@
+  const files = entries
+    .filter((file): boolean => !file.isDirectory())
+    .filter((file): boolean => regex.test(file.name))
+    .map((file): string => `${path}${file.name}`);
+
+  const folders = entries.filter((folder): boolean => folder.isDirectory());
+
+  for (const folder of folders) {
+    files.push(...await getFilePaths(`${path}${folder.name}/`, regex));
```
```suggestion
    .map((file): string => joinFilePath(path, file.name));

  const folders = entries.filter((folder): boolean => folder.isDirectory());

  for (const folder of folders) {
    files.push(...await getFilePaths(joinFilePath(path, folder.name), regex));
```

Reduces the odds of OS differences being an issue.

joachimvh reviewed (changes requested):
Looks good, just some refactor comments.

joachimvh commented on package.json:
```diff
@@ -69,6 +70,12 @@
   },
+  "standard-version": {
+    "scripts": {
+      "postbump": "ts-node ./scripts/upgradeConfig.ts",
```
```suggestion
      "postbump": "npm run upgrade-config",
```
Should prevent having duplicate scripts in general.

MisterTimn replied:
Actually might even just remove the script, originally added it to use the npm env variables. It also won't be executed separately (and if needed one can still do so using ts-node)

joachimvh reviewed (changes requested):
Looks good. Just a very minor comment but then it's ready to merge.
